#249: stargate bugfix and tuning

jakerobb opened the pull request:
**What this PR does**:
Resolving mathematical error in the Stargate pod's memory specs; adjusting memory defaults and also liveness/readiness probes for improved startup reliability.

**Which issue(s) this PR fixes**:
Partially addresses #249 -- the mathematical error is gone. I have been unable to reproduce the issue John sees on his side; Stargate runs quite well for me with a 160MB heap and 320MB memory limit. I also added documentation for the Stargate tuning settings, which was called out in that issue.

**Checklist**
- [x] Changes manually tested
- [ ] Chart versions updated (if necessary)
- [x] Automated Tests added/updated
- [x] Documentation added/updated
- [x] CLA Signed:  [DataStax CLA](https://cla.datastax.com/)

jakerobb commented on charts/k8ssandra/dashboards/stargate.json:
```diff
@@ -0,0 +1,463 @@
+{
```
I'm not sure how or why this commit got included on this PR; I committed it on a different branch as part of #282.

jakerobb commented on charts/k8ssandra/templates/grafana/configmap.yaml:
```diff
@@ -13,4 +13,8 @@ data:
 {{ .Files.Get "dashboards/overview.json" | indent 4 }}
```
This file is also from a commit on a different branch.

bradfordcp commented on charts/k8ssandra/templates/stargate/stargate.yaml:
```diff
@@ -77,16 +77,16 @@ spec:
             limits:
               cpu: {{ .Values.stargate.cpuLimMillicores | default 1000 }}m
-              memory: {{ mul 1.5 (.Values.stargate.heapMB | default 1024) }}Mi
+              memory: {{ mul 4 (.Values.stargate.heapMB | default 256) }}Mi
```
Can we move the default into `values.yaml`?

This removes the need for the commented out documentation line (or even a placeholder which is the style used in the `prometheusSpec` block). My preference is:

```yaml
  prometheus:
    # Provisions an instance of Prometheus as part of this release
    enabled: true
    prometheusSpec:
      # Uncomment to listen for Prometheus traffic under a route prefix
      # externalUrl: http://localhost:9090/prometheus
      # routePrefix: /prometheus
```

but it may make sense to align behind the format used in https://github.com/norwoodj/helm-docs

jsanda replied:
This dovetails into #258.

In general though I do agree that if we are setting defaults we should set them in value.yaml. In some cases it may not be possible though, like if the default is a calculated value.

jsanda replied:
I have been following https://helm.sh/docs/chart_best_practices/values/ as a guide for how to document properties in `values.yaml`. 

What is not entirely clear to me is how best to document properties are commented out by default. Here is what we have in values.yaml for Stargate:

```yaml
stargate:
  # configuration regarding the Stargate API
  enabled: false
  replicas: 1
  clusterVersion: "3.11" # this is the only supported value for now
  containerImage: "auto" # chooses image based on clusterVersion
  # heapMB sets the heap size Stargate will use in megabytes. Default 256. Memory request and limit for the pod will be set to this value x2 and x4, respectively.
  # cpuReqMillicores sets the CPU request for the Stargate pod in millicores. default 200.
  # cpuLimMillicores sets the CPU limit for the Stargate pod in millicores. default 1000.
```
I think it could be easy to miss that those are properties. I think I would prefer to see this:

```yaml
stargate:
  # configuration regarding the Stargate API
  enabled: false
  replicas: 1
  clusterVersion: "3.11" # this is the only supported value for now
  containerImage: "auto" # chooses image based on clusterVersion
  # heapMB sets the heap size Stargate will use in megabytes. Default 256. Memory request and limit for the pod will be set to this value x2 and x4, respectively.
  # heapMB:
  
  # cpuReqMillicores sets the CPU request for the Stargate pod in millicores. default 200.
  # cpuReqMillicores:

  # cpuLimMillicores sets the CPU limit for the Stargate pod in millicores. default 1000.
  # cpuLimMillicores:
```
I realize that this is likely to highly subjective :)

jakerobb replied:
Makes sense to me, John. Chris's example includes reasonable values (possibly matching the default?) on the commented out lines. I think I like that better than leaving it blank. Thoughts?

jsanda replied:
If we have default values that we are going to use, then yes definitely set them in `values.yaml`.

bradfordcp commented on charts/k8ssandra/values.yaml:
```diff
@@ -87,6 +87,9 @@ stargate:
   replicas: 1
   clusterVersion: "3.11" # this is the only supported value for now
   containerImage: "auto" # chooses image based on clusterVersion
+  # heapMB sets the heap size Stargate will use in megabytes. Default 256. Memory request and limit for the pod will be set to this value x2 and x4, respectively.
```
See comment on `charts/k8ssandra/templates/stargate/stargate.yaml`

jsanda commented on charts/k8ssandra/templates/stargate/stargate.yaml:
```diff
@@ -77,16 +77,16 @@ spec:
             requests:
               cpu: {{ .Values.stargate.cpuReqMillicores | default 200 }}m
-              memory: {{ mul 1.5 (.Values.stargate.heapMB | default 1024) }}Mi
+              memory: {{ mul 2 (.Values.stargate.heapMB | default 256) }}Mi
```
Can we put this in a variable to eliminate the duplication?

jsanda commented on charts/k8ssandra/templates/stargate/stargate.yaml:
```diff
@@ -77,16 +77,16 @@ spec:
-                -Xms{{ .Values.stargate.heapMB | default 1024 }}M
-                -Xmx{{ .Values.stargate.heapMB | default 1024 }}M
+                -Xms{{ .Values.stargate.heapMB | default 256 }}m
+                -Xmx{{ .Values.stargate.heapMB | default 256 }}m
```
Can we put this in a variable to eliminate the duplication?

jsanda commented on charts/k8ssandra/templates/stargate/stargate.yaml:
```diff
@@ -77,16 +77,16 @@ spec:
           resources:
```
I think it makes sense to make `resources` in general configurable. See [here](https://github.com/k8ssandra/k8ssandra/blob/main/charts/k8ssandra/templates/cassandra/cassdc.yaml#L53) for an example. 

Given that we have some insights into memory configuration I would be inclined to do as follows:

* If no values are set the apply defaults as you currently do
* else if heap/memory is specified, use those overrides
* else if `resources` is specified, use that.

The user directly setting `resources` would be a more advanced scenario IMO but one that we definitely want to support.

jsanda reviewed (changes requested):
@jakerobb overall the changes look good!

Can you include a fix for #147 in this PR as well?

I tested your branch and Stargate was stable with the default settings. I did not however make any client requests? Did you test with any load?

jakerobb commented:
> Can you include a fix for #147 in this PR as well?

Will do.

> I tested your branch and Stargate was stable with the default settings. I did not however make any client requests? Did you test with any load?

Yes, I threw one million requests at two Stargate nodes over the course of three hours. The screenshot of the dashboard on #282 shows a shorter duration of a similar test.

jsanda commented:
>Yes, I threw one million requests at two Stargate nodes over the course of three hours. The screenshot of the dashboard on #282 shows a shorter duration (I think it was 100k requests) of a similar test running against an even smaller heap.

Awesome!

jsanda reviewed (changes requested):
Looks like we still need to address #147. I still see the `clusterVersion` and `containerImage` properties.

When you replace the `containerImage` property with `image` also be sure to add an `imagePullPolicy` property as well.

jakerobb commented:
Okay, I have:
1. Rebased on main
2. Moved defaults from template to values.yaml
3. documented options in values.yaml
4. removed stargate.clusterVersion property
5. changed behavior of stargate.image as described on #147
6. added stargate.imagePullPolicy
7. created a `utils` package and moved some utility functions there
8. added several unit tests for Stargate
9. updated our test suite to use unique, descriptive values for namespace, clusterName, and releaseName.

Re: #7 -- I initially planned to create a single "utils.go" file, but instead decided to separate them by subject (EnvVars, Containers, ...). This minimizes the chances for any one file to get big and difficult to manage, reduces namespace pollution when importing, and feels like a straightforward way to divide them up.

Re: #8 -- more tests are coming in separate PRs.

Re: #9 -- I'll be adding a test that, paired with this change, would have caught #293.

Tests are passing; I think this is good to merge. @bradfordcp and @jsanda, here's a poke because you guys requested changes.

jsanda commented on tests/unit/utils/containers.go:
```diff
@@ -0,0 +1,48 @@
+	coreV1 "k8s.io/api/core/v1"
+)
+
+func GetInitContainer(cassdc *cassdcV1beta1.CassandraDatacenter, name string) *coreV1.Container {
```
Can you change the name to `GetCassdcInitContainer`? You cannot have overloaded functions in Go :(

jsanda commented on tests/unit/utils/containers.go:
```diff
@@ -0,0 +1,48 @@
+
+}
+
+func GetContainer(cassdc *cassdcV1beta1.CassandraDatacenter, name string) *coreV1.Container {
```
Can you change the function name to `GetCassdcContainer`?

jsanda commented on tests/unit/utils/containers.go:
```diff
@@ -0,0 +1,48 @@
+
+}
+
+func AssertInitContainerNamesMatch(cassdc *cassdcV1beta1.CassandraDatacenter, names ...string) {
```
I was just thinking about my previous two comment. Maybe it makes more sense to put CassandraDatacenter-specific function into their own package, e.g., `cassdc`. Then we can just write `cassdc.GetContainer`. vs `GetCassdcContainer`? 

wdty?

Note that I am using `cassdc` because it is the abbreviation specified in the CassandraDatacenter CRD. It allows to me to do `kubectl get cassdc` for example.

jakerobb replied:
I like the idea of having the functions in a separate package. It improves the focus of the package, shortens function names without losing clarity, and in places where you only need one, you can dot-alias the imports and not qualify them at all.

I made the change locally to move the cassdc functions to a separate package, but then found that `cassdc` is a very common (and reasonable) variable name, and using `cassdc` as the import alias creates a conflict. The conflict doesn't cause a compile failure; the variables just win. I conclude that `cassdc` is not a good choice for the alias. I changed it to use a dot alias in template_cassdc_test, and I think I like the result even more. :)

jsanda commented on tests/unit/utils/envVars.go:
```diff
@@ -0,0 +1,14 @@
+	coreV1 "k8s.io/api/core/v1"
+)
+
+func FindEnvVarByName(haystack []coreV1.EnvVar, needle string) *coreV1.EnvVar {
```
While this made me laugh, it might be confusing to other users in the future. Maybe change these to something like `envVars` and `name` or `envVarName`.

jakerobb replied:
Heh, that was a leftover from when I was trying to write it generically. 

FWIW, I've used this nomenclature in the past (in Java), and never had it confuse anyone. It's both whimsical and very obvious. :)

Nonetheless, since it's not generic in any way, it makes sense to change it here.

jakerobb commented on charts/k8ssandra/templates/stargate/stargate.yaml:
```diff
@@ -8,7 +10,7 @@ metadata:
 {{ include "k8ssandra.labels" . | indent 4 }}
 spec:
-  replicas: {{ default 1 .Values.stargate.replicas }}
+  replicas: {{ required "stargate.replicas is required" .Values.stargate.replicas }}
```
I thought it made sense to keep it in order to handle gracefully the scenario where the user overrides the value to empty (presumably inadvertently). Ditto for the other `required`s I added in this file.

jsanda commented on tests/unit/template_stargate_test.go:
```diff
@@ -0,0 +1,186 @@
+	. "github.com/k8ssandra/k8ssandra/tests/unit/utils"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	appsV1 "k8s.io/api/apps/v1"
```
minor nit - can you make the alias all lower case. That seems to be the common convention.

jsanda commented on tests/unit/utils/containers.go:
```diff
@@ -0,0 +1,48 @@
+package utils
+
+import (
+	cassdcV1beta1 "github.com/datastax/cass-operator/operator/pkg/apis/cassandra/v1beta1"
```
minor nit - can you make the aliasas all lower case. That seems to be the common convention.

jsanda commented on tests/unit/utils/envVars.go:
```diff
@@ -0,0 +1,14 @@
+package utils
+
+import (
+	coreV1 "k8s.io/api/core/v1"
```
minor nit - can you make the alias all lower case. That seems to be the common convention.

jsanda replied:
minor nit - can you make the alias all lower case. That seems to be the common convention.

jsanda commented on tests/unit/utils/envVars.go:
```diff
@@ -0,0 +1,14 @@
+package utils
```
Can you rename the file to either `envvars.go` or `env_vars.go` to be consistent with naming conventions.

jsanda reviewed (changes requested):
Can you also bump the k8ssandra chart version.

jsanda commented on tests/unit/utils/cassdc/containers.go:
```diff
@@ -0,0 +1,32 @@
+package cassdc
+
```
I like the reorganization! Just so you are aware a lot of these functions will likely be needed in reaper-operator and in medusa-operator. At some point soon we can pull the common test util code out into its own project.

Since I initially created these functions, I can add a follow up PR with some docs after yours is merged.

jsanda commented on tests/unit/utils/kubeapi/ingress.go:
```diff
@@ -0,0 +1,42 @@
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+func FindIngressRuleByHttpPath(rules []networkingV1.IngressRule, path string) (*networkingV1.IngressRule, *networkingV1.HTTPIngressPath) {
```
Can you add some comments for this?

jsanda commented on tests/unit/utils/kubeapi/ingress.go:
```diff
@@ -0,0 +1,42 @@
+	return nil, nil
+}
+
+func VerifyNoRuleWithPath(rules []networkingV1.IngressRule, path string) {
```
Can you add some comments for this?

jsanda commented on tests/unit/utils/kubeapi/ingress.go:
```diff
@@ -0,0 +1,42 @@
+	ExpectWithOffset(1, authRule).To(BeNil())
+}
+
+func VerifyIngressRule(rules []networkingV1.IngressRule, releaseName, path string, pathType *networkingV1.PathType, host string, port int) {
```
Can you add some comments for this?

jsanda commented on tests/unit/utils/traefik/httpRoutes.go:
```diff
@@ -0,0 +1,39 @@
+	traefik "github.com/traefik/traefik/v2/pkg/provider/kubernetes/crd/traefik/v1alpha1"
+)
+
+func VerifyTraefikHTTPIngressRoute(ingress traefik.IngressRoute, entrypoint string, match string, backendService string, backendPort int) {
```
Can you add some comments for this?

jsanda commented on tests/unit/utils/traefik/httpRoutes.go:
```diff
@@ -0,0 +1,39 @@
+	ExpectWithOffset(1, service.Port).To(Equal(int32(backendPort)))
+}
+
+func findTraefikHTTPRouteByMatch(spec *traefik.IngressRouteSpec, match string) *traefik.Route {
```
Can you add some comments for this?

jsanda commented on tests/unit/utils/traefik/tcpRoutes.go:
```diff
@@ -0,0 +1,39 @@
+	traefik "github.com/traefik/traefik/v2/pkg/provider/kubernetes/crd/traefik/v1alpha1"
+)
+
+func VerifyTraefikTCPIngressRoute(ingress traefik.IngressRouteTCP, entrypoint string, match string, backendService string, backendPort int) {
```
Can you add some comments for this?

jsanda reviewed (changes requested):
I made an initial pass through the first few commits. I will do another review later this afternoon.

jakerobb commented:
I've rebased this branch with the latest from main. Merging was fun! I've also added the requested doc comments, removed some debugging output, and taken care of a couple bits of fallout from the rebase.

jakerobb commented:
Rebasing has added dozens of unrelated commits to this PR. I think I did something wrong.

I am going to create a new branch, cherry pick the commits, and create a new PR. Leaving this open for now so that we don't lose track of the feedback and conversations in flight.

jakerobb commented:
This is now superseded by PR #333, which contains mostly the same changes but without all of the unrelated commits.
